Cubical/HITs/Modulo/Properties.agda

```diff
@@ -0,0 +1,71 @@
+{-# OPTIONS --cubical --safe #-}
+
+module Cubical.HITs.Modulo.Properties where
+
+open import Cubical.Core.Everything
+
+open import Cubical.Data.Nat
+open import Cubical.Data.Fin
+
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.Isomorphism
+
+open import Cubical.HITs.Modulo.Base
+
+private
+  variable
+    ℓ : Level
+    k : ℕ
+
+Modulo0≡ℕ : Modulo 0 ≡ ℕ
+Modulo0≡ℕ = isoToPath lemma
+ where
+ open Iso
+ lemma : Iso (Modulo 0) ℕ
+ fun lemma (embed n) = n
+ fun lemma (step n i) = n
+ fun lemma (squash z n i j) = n
+ inv lemma = embed
+ rightInv lemma n = refl
+ leftInv lemma (embed n) = refl
+ leftInv lemma (step n i) j = squash _ n j i
+ leftInv lemma (squash z n i j) k = squash z n (i ∧ k) j
+
+Modulo0-isSet : isSet (Modulo 0)
+Modulo0-isSet = subst isSet (sym Modulo0≡ℕ) isSetℕ
+
+isContrModulo1 : isContr (Modulo 1)
+isContrModulo1
+  = value , λ
+  { (embed n) → universal n
+  ; (step n i) → universal-path n i
+  ; (squash () n i j)
+  }
+  where
+  value : Modulo 1
+  value = embed 0
+
+  universal : ∀ n → value ≡ embed n
+  universal zero = refl
+  universal (suc n) = universal n ∙ step n
+
+  universal-path
+   : ∀ n → PathP (λ i → value ≡ step n i) (universal n) (universal (suc n))
+  universal-path n i j
+    = compPath-filler (universal n) (step n) i j
+
+steps : ∀ m n → embed {k = k} m ≡ embed (n * k + m)
+steps m zero = refl
+steps {k} m (suc n)
+  = steps m n ∙ step (n * k + m) ∙ cong embed (+-assoc k (n * k) m)
+
+fembed : Fin k → Modulo k
+fembed (n , n<k) = embed n
+
+MReduction : (k : ℕ) → Modulo (suc k) → Set
+MReduction k (embed n) = Reduction (suc k) n
+MReduction k (step n i) = Reduction≡ (suc k) n (suc k + n) refl i
+
+mreduce : (m : Modulo (suc k)) → MReduction k m
+mreduce (embed n) = reduce _ n
+mreduce {k} (step n i) = reduce≡ k n (suc k + n) refl i
```

Comment:
Can you prove that `Fin k` is equivalent to `Modulo k` ?

Reply:
I guess I didn't get as far as I wanted with just this, because the reduction function started consuming all my effort (and then I forgot about the other stuff). However, there are some interesting differences between `Fin` and `Modulo`.

- `Fin 0` is empty and `Modulo 0` is the natural numbers
- It's easy to have a successor function `Fin k -> Fin (suc k)`, and harder to have one typed `Fin k -> Fin k`. However, the case is the opposite for `Modulo`.
- I think this probably generalizes somewhat, and it's easy to write arithmetic internal to `Modulo k`, and the difficulty related to writing it in `Fin k` is the reduction function (although I suppose I've factored that out in this).

I can add that `Fin (suc k)` is equivalent to `Modulo (suc k)`.

Reply:
This is interesting! Maybe you could include some of this comment in the top of the Modulo file to explain why it's a nice alternative to Fin? I'm very interested in programming examples that get simpler with univalence and HITs.

Reply:
And yes, please prove that the two types are equivalent. That would be nice to have as well as we can transport results between the two representations. I wonder if there is anything that is really painful with `Fin` that gets a lot simpler with `Modulo`....

Reply:
`Modulo (suc k)` and `Fin (suc k)` have been shown equivalent.

Reply:
By the way, I don't know if I mentioned this elsewhere, but I changed the definition of `Modulo` along the way. Instead of trying to truncate the 0 case, I avoid adding paths to it. This made things rather simpler, and a pattern synonym made it so that the successor cases weren't really any different.